ui/raidboss/raidemulator.js

```diff
@@ -0,0 +1,832 @@
+    // 1A = gaining effects -> 1E losing effects | 1B = head Markers | 01 = zone change
+    // 03,04 = adding/removing combatant | 21 = wipe?
+    // Add more to regex only if needed
+    fight.logs = fight.logs.filter(function(line) {
+      return / 1[456AB]:| 0[134]:| 00:003[89]:| 21:.{8}:/.test(line);
+    }).sort();
+
+    this.fights.push(fight);
+    if (this.addFightCallback)
+      this.addFightCallback(fight);
+  }
+
+  IsWipe(log) {
+    // Actor control line list: https://gist.github.com/quisquous/250001cbce232a48e6a9ce772a56675a
+    return log.match(/ 21:........:40000010:/) || log.indexOf('00:0038:cactbot wipe') != -1;
+  }
+
+  ParseZoneChange(log) {
+    let m = log.match(/ 01:Changed Zone to (.*)\./);
+    if (!m)
+      return;
+    return m[1];
+  }
+}
+
+// Responsible for playing back a fight and emitting events as needed.
+class LogPlayer {
+  LogPlayer() {
+    this.Reset();
+  }
+
+  Reset() {
+    this.fight = null;
+    this.localStartTime = null;
+    this.logIdx = null;
+  }
+
+  SendLogEvent(logs) {
+    let evt = new CustomEvent('onLogEvent', { detail: { logs: logs } });
+    document.dispatchEvent(evt);
+  }
+
+  SendZoneEvent(zoneName) {
+    let evt = new CustomEvent('onZoneChangedEvent', { detail: { zoneName: zoneName } });
+    document.dispatchEvent(evt);
+  }
+
+  SendPlayerEvent(name, job, id) {
+    let jobDetail = {
+      oath: 100,
+      beast: 100,
+      blood: 100,
+      lillies: 3,
+      aetherflowStacks: 3,
+      fairyGauge: 100,
+      drawMilliseconds: 15000,
+      drawnCard: 'Speer',
+      spreadCard: 'Baum',
+      roadCard: 'Waage',
+      arcanumCard: 'König der Kronen',
+      lightningStacks: 3,
+      lightningMilliseconds: 12000,
+      chakraStacks: 5,
+      bloodMilliseconds: 10000,
+      lifeMilliseconds: 10000,
+      eyesAmount: 3,
+      hutonMilliseconds: 10000,
+      ninkiAmount: 3,
+      samuraiGauge: 100,
+      songName: 'Mages Ballad',
+      songProcs: 3,
+      overheatMilliseconds: 10000,
+      heat: 100,
+      ammunition: 3,
+      gauss: 'gauss',
+      umbralStacks: 3,
+      umbralMilliseconds: 10000,
+      umbralHearts: 3,
+      enochian: 'enochian',
+      polygot: 'polygot',
+      nextPolygotMilliseconds: 10000,
+      dreadwyrmStacks: 3,
+      bahamutStacks: 2,
+      dreadwyrmMilliseconds: 10000,
+      bahamutMilliseconds: 10000,
+      whiteMana: 100,
+      blackMana: 100,
+    };
+    let evt = new CustomEvent('onPlayerChangedEvent', { detail: {
+      id: id,
+      name: name,
+      job: job,
+      currentHP: 1000,
+      maxHP: 1000,
+      currentMP: 1000,
+      maxMP: 1000,
+      currentTP: 1000,
+      maxTP: 1000,
+      currentCP: 1000,
+      maxCP: 1000,
+      currentGP: 1000,
+      maxGP: 1000,
+      level: 100,
+      debugJob: 'debug',
+      jobDetail: jobDetail,
+      pos: { x: 0, y: 0, z: 0 },
+    } });
+    document.dispatchEvent(evt);
+    document.getElementById('player').textContent = name;

```

Comment:
From an abstraction point of view, it'd be nice if the LogPlayer didn't mess with the dom at all and the EmulatorView did all of that.  That's part of what I was trying to get at with reorganizing here, i.e. a LogCollector that only just handled importing logs and breaking them up, a LogPlayer that handles interfacing with the underlying raidboss, sending events, handling time, and finally an EmulatorView that messes with the dom and shows things.